enterprise/backend/src/metabase_enterprise/data_editing/models/undo.clj

```diff
@@ -0,0 +1,251 @@
+  matching the given where clause. Returns 0 if we do not need to prune anything."
+  [batches-to-keep & [where]]
+  (-> {:select   [:batch_num]
+       :from     [(t2/table-name :model/Undo)]
+       :where    (or where true)
+       :group-by :batch_num
+       :order-by [[:batch_num :desc]]
+       :limit    1
+       :offset   batches-to-keep}
+      t2/query
+      first
+      :batch_num
+      (or 0)))
+
+(defn- batch-to-prune-from-for-rows
+  "This is like [[batch-to-prune-from], except that we're enforcing a max on the row count."
+  [rows-to-keep & [where]]
+  (-> {:select   [:batch_num]
+       :from     [(t2/table-name :model/Undo)]
+       :where    (or where true)
+       :order-by [[:id :desc]]
+       :limit    1
+       :offset   rows-to-keep}
+      t2/query
+      first
+      :batch_num
+      (or 0)))
+
+(defn- prune-from-batch! [batch-num & [where]]
+  (t2/delete! :model/Undo
+              :batch_num [:<= batch-num]
+              {:where (or where true)}))
+
+(defn- prune-batches! [batches-to-keep & [where]]
+  (prune-from-batch! (batch-to-prune-from batches-to-keep where) where))
+
+(defn track-change!
+  "Insert some snapshot data based on edits made to the given table."
+  [user-id scope table-id->row-pk->values]
+  (let [scope (serialize-scope scope)]
+    (t2/with-transaction [_conn]
+      (let [seq-name       "undo_batch_num"
+            next-batch-num (or (t2/select-one-fn :next_val [:sequences :next_val] :name seq-name {:for :update}) 1)]
+        (t2/update! :sequences {:name seq-name} {:next_val (inc next-batch-num)})
+        (t2/insert!
+         :model/Undo
+         (for [[table-id table-updates] table-id->row-pk->values
+               [row-pk values] table-updates]
+           (merge {:batch_num  next-batch-num
+                   :table_id   table-id
+                   :user_id    user-id
+                   :row_pk     row-pk
+                   :scope      scope
+                   ;; default value, can be override in values
+                   :undoable   true}
+                  values)))))
+
+    ;; Delete snapshots that have been undone, as we keep a linear history and will no longer be able to "redo" them.
+    (when-let [{:keys [batch_num]} (first (next-batch false user-id scope))]
+      (t2/delete! :model/Undo
+                  :batch_num [:>= batch_num]
+                  :scope scope
+                  :undone true))
+
+    ;; Pruning. Fairly naive implementation. Doesn't assume we were fully pruned before this update.
+
+    (prune-from-batch! (max (batch-to-prune-from-for-rows retention-total-rows)
+                            (batch-to-prune-from retention-total-batches)))
+
+    (let [where-scope [:= :scope scope]]
+      (prune-batches! retention-batches-per-scope where-scope))
+
+    (let [where-user [:= :user_id user-id]]
+      (prune-batches! retention-batches-per-user where-user))))
```

Comment:
I can't remember why I used let blocks here, feels silly. Let's inline the clauses.